libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,149 @@
+	solAssert(m_currentNode == nullptr, "Current node not properly reset.");
+
+	return graph;
+}
+
+bool FunctionCallGraphBuilder::visit(Identifier const& _identifier)
+{
+	if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_identifier.annotation().referencedDeclaration))
+	{
+		if (*_identifier.annotation().requiredLookup == VirtualLookup::Virtual)
+			callable = &callable->resolveVirtual(*m_contract);
+
+		solAssert(
+			*_identifier.annotation().requiredLookup != VirtualLookup::Super,
+			"Only MemberAccess can have lookup 'super'"
+		);
+
+		// Ignore calls to other contract
+		if (callable->annotation().contract && !m_contract->derivesFrom(*callable->annotation().contract) && !callable->annotation().contract->isLibrary())
+			return true;
+
+		if (!m_nodes->count(callable))
+			visitCallable(callable);
+
+		m_currentNode->calls.emplace(*m_nodes->find(callable));
+	}
+
+	return true;
+}
+
+bool FunctionCallGraphBuilder::visit(NewExpression const& _newExpression)
+{
+	if (ContractType const* contractType = dynamic_cast<ContractType const*>(_newExpression.typeName().annotation().type))
+	{
+		CallableDeclaration const* callable = contractType->contractDefinition().constructor();
+
+		if (!m_nodes->count(callable))
+			m_nodes->emplace(make_shared<Node>(callable));
+	}
+
+	return true;
+}
+
+bool FunctionCallGraphBuilder::visit(MemberAccess const& _memberAccess)
+{
+	if (auto const* callable = dynamic_cast<CallableDeclaration const*>(_memberAccess.annotation().referencedDeclaration))
+	{
+		if (*_memberAccess.annotation().requiredLookup == VirtualLookup::Super)
+		{
+			ContractDefinition const* super =
+				m_currentNode->callable ?
+				m_currentNode->callable->annotation().contract :
+				m_contract;
+
+			callable = &callable->resolveVirtual(*m_contract, super);
+		}
+
+		solAssert(
+			*_memberAccess.annotation().requiredLookup != VirtualLookup::Virtual,
+			"MemberAccess cannot have lookup 'virtual'"
+		);
+
+		// Ignore calls to other contracts
+		if (callable->annotation().contract && !m_contract->derivesFrom(*callable->annotation().contract) && !callable->annotation().contract->isLibrary())
+			return true;
+
+		if (!m_nodes->count(callable))
+			visitCallable(callable);
+	}
+
+	return true;
+}
+
+void FunctionCallGraphBuilder::visitCallable(CallableDeclaration const* _callable)
+{
+	solAssert(!m_nodes->count(_callable), "");
+
+	auto previousCallable = m_currentNode;
+	m_nodes->emplace(m_currentNode = make_shared<Node>(_callable));
+	_callable->accept(*this);
+	m_currentNode = previousCallable;
+}
+
+void FunctionCallGraphBuilder::visitConstructor(ContractDefinition const& _contract)
+{
+	auto previousNode = m_currentNode;
+	m_nodes->emplace(m_currentNode = make_shared<Node>(_contract.constructor()));
+	if (previousNode)
+		previousNode->calls.emplace(m_currentNode);
+
+	if (_contract.annotation().linearizedBaseContracts.size() > 1)
+		visitConstructor(*_contract.annotation().linearizedBaseContracts[1]);
```

Comment:
I think you should visit all constructors of all linearized base contracts of the current most derived contract. Otherwise you miss some bases in diamond inheritance, don't you?

Reply:
Ah, I see what you mean. Fair point.